2015/presentations/import_deep_dive/text.md

```diff
@@ -0,0 +1,297 @@
+
+If you're interested in this, I'll just direct you to
+[importlib documentation](https://docs.python.org/3/library/importlib.html).
+Now that you know the terminology and and the "normal" way of how things are
+done, those docs should make much more sense.
+
+Instead, I'll mention some real-world projects that use custom import hooks,
+which you can use as inspiration about what's possible.
+
+Hy (hy.readthedocs.org) allows you to import modules written in a dialect of
+Lisp. Or as its docs put it, it's "lisp-stick on a Python".
+It uses a custom finder that looks for `*.hy` files instead of `*.py` ones,
+and a corresponding loader that parses the Lisp code, runs it, and populates
+a Python module object with the resulting functions and objects.
+
+This is a good approach for any project that wants to embed a custom language
+in the Python runtime.
+
+The Macropy project uses a similar technique to add syntactic macros to
+Python code. Conceptually, it's the same as Hy, except the language is a bit
+more similar to Python.
+
+Perhaps most widely used of import customizations is Cython's `pyximport`,
+which finds a `.pyx` file, compiles it into a C extension, and then imports it.
+
+If you just want to load Python modules that are stored in a non-standard
+location, like a database, Git repository, or just a string,
+look near the end of David Beazley's PyCon US 2015 tutorial for his
+Redis importer.
+
+
+## Loading Module Resources
+
+Every module uses a `Loader` object when it is first initialized, but the
+usefulness of loaders doesn't end there.
+
+Loaders of most types of modules implement the `ResourceLoader` ABC, a protocol
+that allows loading data files stored alongside a module.
+For example, to open a data file stored "next to" the current module on the
+filesystem, a common approach is to do:
+
+    >>> import os
+    >>> filename = os.path.join(os.path.dirname(__file__), "file.dat")
+    >>> with open(filename, 'rb') as f:
+    ...     data = f.read()
+
+However, this will only work when the module and data is loaded from regular
+files. If they were, for example, distributed in a zip file, the `open`
+function would fail.
+
+A more robust approach is to do:
+
+    >>> import os
+    >>> filename = os.path.join(os.path.dirname(__file__), "file.dat")
+    >>> data = __loader__.get_data(filename)
+
+While constructing the filename is still somewhat painful, this approach will
+work with zipped packages, and any other loaders that implement `get_data`.
+
+(Sadly, David Beazley's Redis importer does not yet support this.)
+
+
+## Extension and built-in modules
+
+A big part of CPython's appeal (especially back when it was just starting) is
+that it is easily extensible with code written in C, so you can interface with
+existing libraries or write code that runs fast.
+Today, it's generally better to use CFFI or Cython for this, but extensions
+remain important. If nothing else, a big part of the the standard library
+is written in C.
```

Comment:
Actually that's not necessarily true. A lot of modules were reimplemented in pure python in CPython.